crates/oxc_linter/src/rules/nextjs/no_before_interactive_script_outside_document.rs

```diff
@@ -0,0 +1,394 @@
+use oxc_ast::{
+    ast::{JSXAttributeItem, JSXAttributeName, JSXAttributeValue, JSXElementName, JSXIdentifier},
+    AstKind,
+};
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::Error,
+};
+use oxc_macros::declare_oxc_lint;
+use oxc_span::Span;
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("eslint-plugin-next(no-before-interactive-script-outside-document): next/script's `beforeInteractive` strategy should not be used outside of `pages/_document.js`")]
+#[diagnostic(
+    severity(warning),
+    help("See https://nextjs.org/docs/messages/no-before-interactive-script-outside-document")
+)]
+struct NoBeforeInteractiveScriptOutsideDocumentDiagnostic(#[label] pub Span);
+
+#[derive(Debug, Default, Clone)]
+pub struct NoBeforeInteractiveScriptOutsideDocument;
+
+declare_oxc_lint!(
+    /// ### What it does
+    /// Prevent usage of `next/script`'s `beforeInteractive` strategy outside of `pages/_document.js`.
+    ///
+    /// ### Why is this bad?
+    ///
+    ///
+    /// ### Example
+    /// ```javascript
+    /// ```
+    NoBeforeInteractiveScriptOutsideDocument,
+    correctness
+);
+
+impl Rule for NoBeforeInteractiveScriptOutsideDocument {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        if let AstKind::JSXOpeningElement(jsx_el) = node.kind() {
+            let Some(path) = ctx.file_path().to_str() else { return };
+            let is_in_app_dir = path.contains("app/") || path.contains("app\\");
+            if is_in_app_dir {
+                return;
+            }
+
+            let Some(page) = path.split("pages").last() else { return };
+            if page.starts_with("/_document") || page.starts_with("\\_document") {
+                return;
+            }
```

Comment:
Again, these expensive should be ordered from heaviest to cheapest.

Btw, is there no common utils for app and page dir checks in other places? (I haven't looked).

Reply:
> Again, these expensive should be ordered from heaviest to cheapest.

I moved document page check to before import name check. let me know if you want me to move it to elsewhere

> Btw, is there no common utils for app and page dir checks in other places? (I haven't looked).

I created an util file for nextjs, and put some of the functions to util.